# Introduction

This is the Northwind Database. Northwind Trades is a fictious company that imports and exports specialty foods around the world. I am going to use sql to analyze this database and find trends, insights and other information about this database.

In [11]:
#Import libraries to run sqlite
import sqlite3
import pandas as pd
pd.set_option('display.max_rows', 50)
#User defined function to run queries more easily
def ex_query(query):
    cnx = sqlite3.connect("C:\\Users\\wally\\Downloads\\Northwind_large.sqlite\\Northwind_large.sqlite")
    table = pd.read_sql_query(query, cnx)
    return table

In [12]:
# List of all table in our Northwind Large database
tables = ex_query("""SELECT name 
                       FROM sqlite_schema 
                      WHERE type ='table' AND name NOT LIKE 'sqlite_%'""")
print(tables)

                    name
0               Employee
1               Category
2               Customer
3                Shipper
4               Supplier
5                 Orders
6                Product
7            OrderDetail
8   CustomerCustomerDemo
9    CustomerDemographic
10                Region
11             Territory
12     EmployeeTerritory


Now that we know how many tables we have, it would be a good idea to get a look at each of them.

In [13]:
#Print employee table
employee = ex_query("""SELECT * 
                         FROM Employee""")
print(employee)

   Id   LastName FirstName                     Title TitleOfCourtesy  \
0   1    Davolio     Nancy      Sales Representative             Ms.   
1   2     Fuller    Andrew     Vice President, Sales             Dr.   
2   3  Leverling     Janet      Sales Representative             Ms.   
3   4    Peacock  Margaret      Sales Representative            Mrs.   
4   5   Buchanan    Steven             Sales Manager             Mr.   
5   6     Suyama   Michael      Sales Representative             Mr.   
6   7       King    Robert      Sales Representative             Mr.   
7   8   Callahan     Laura  Inside Sales Coordinator             Ms.   
8   9  Dodsworth      Anne      Sales Representative             Ms.   

    BirthDate    HireDate                        Address      City  \
0  1980-12-08  2024-05-01     507 - 20th Ave. E. Apt. 2A   Seattle   
1  1984-02-19  2024-08-14             908 W. Capital Way    Tacoma   
2  1995-08-30  2024-04-01             722 Moss Bay Blvd.  Kirkland   

In [14]:
#Print category table
category = ex_query("""SELECT * 
                         FROM category""")
print(category)

   Id    CategoryName                                        Description
0   1       Beverages        Soft drinks, coffees, teas, beers, and ales
1   2      Condiments  Sweet and savory sauces, relishes, spreads, an...
2   3     Confections                Desserts, candies, and sweet breads
3   4  Dairy Products                                            Cheeses
4   5  Grains/Cereals                Breads, crackers, pasta, and cereal
5   6    Meat/Poultry                                     Prepared meats
6   7         Produce                          Dried fruit and bean curd
7   8         Seafood                                   Seaweed and fish


In [15]:
#Print customer table
customer = ex_query("""SELECT * 
                         FROM customer""")
print(customer)

       Id                         CompanyName              ContactName  \
0   ALFKI                 Alfreds Futterkiste             Maria Anders   
1   ANATR  Ana Trujillo Emparedados y helados             Ana Trujillo   
2   ANTON             Antonio Moreno Taquería           Antonio Moreno   
3   AROUT                     Around the Horn             Thomas Hardy   
4   BERGS                  Berglunds snabbköp       Christina Berglund   
..    ...                                 ...                      ...   
86  WARTH                      Wartian Herkku         Pirkko Koskitalo   
87  WELLI              Wellington Importadora            Paula Parente   
88  WHITC                White Clover Markets           Karl Jablonski   
89  WILMK                         Wilman Kala          Matti Karttunen   
90  WOLZA                      Wolski  Zajazd  Zbyszek Piestrzeniewicz   

                 ContactTitle                        Address         City  \
0        Sales Representative     

In [16]:
#Print shipper table
shipper = ex_query("""SELECT *
                        FROM shipper""")
print(shipper)

   Id       CompanyName           Phone
0   1    Speedy Express  (503) 555-9831
1   2    United Package  (503) 555-3199
2   3  Federal Shipping  (503) 555-9931


In [17]:
#Print supplier table
supplier = ex_query("""SELECT *
                        FROM supplier""")
print(supplier)

    Id                             CompanyName                 ContactName  \
0    1                          Exotic Liquids            Charlotte Cooper   
1    2              New Orleans Cajun Delights               Shelley Burke   
2    3               Grandma Kelly's Homestead               Regina Murphy   
3    4                           Tokyo Traders                Yoshi Nagase   
4    5      Cooperativa de Quesos 'Las Cabras'  Antonio del Valle Saavedra   
5    6                                Mayumi's                 Mayumi Ohno   
6    7                           Pavlova, Ltd.                 Ian Devling   
7    8                Specialty Biscuits, Ltd.                Peter Wilson   
8    9                        PB Knäckebröd AB               Lars Peterson   
9   10               Refrescos Americanas LTDA                 Carlos Diaz   
10  11             Heli Süßwaren GmbH & Co. KG               Petra Winkler   
11  12       Plutzer Lebensmittelgroßmärkte AG                 M

Order is a keyword and therefore needed to be enclosed by backticks to work properly.
So now we can rename the table 'orders' to avoid this problem.

In [18]:
#Print order table
order = ex_query("""SELECT *
                      FROM orders""")
print(order)


          Id CustomerId  EmployeeId            OrderDate         RequiredDate  \
0      10248      VINET           5           2012-07-04           2012-08-01   
1      10249      TOMSP           6           2012-07-05           2012-08-16   
2      10250      HANAR           4           2012-07-08           2012-08-05   
3      10251      VICTE           3           2012-07-08           2012-08-05   
4      10252      SUPRD           4           2012-07-09           2012-08-06   
...      ...        ...         ...                  ...                  ...   
16813  27061      FOLKO           5  2013-06-29 21:05:55  2013-08-02 04:10:53   
16814  27062      FRANK           2  2014-01-19 12:27:11  2014-01-24 15:15:31   
16815  27063      ALFKI           5  2014-10-15 09:51:09  2014-11-11 14:31:37   
16816  27064      TRADH           8  2013-02-07 02:06:05  2013-03-14 09:43:16   
16817  27065      ANATR           1  2013-08-31 02:59:28  2013-09-15 23:11:49   

               ShippedDate 

In [19]:
#Change order table to 'Orders'

# ex_query("""ALTER TABLE `order`
# RENAME TO Orders""")

In [20]:
#Print product table
product = ex_query("""SELECT *
                      FROM product""")
print(product)

    Id                      ProductName  SupplierId  CategoryId  \
0    1                             Chai           1           1   
1    2                            Chang           1           1   
2    3                    Aniseed Syrup           1           2   
3    4     Chef Anton's Cajun Seasoning           2           2   
4    5           Chef Anton's Gumbo Mix           2           2   
..  ..                              ...         ...         ...   
72  73                       Röd Kaviar          17           8   
73  74                    Longlife Tofu           4           7   
74  75             Rhönbräu Klosterbier          12           1   
75  76                     Lakkalikööri          23           1   
76  77  Original Frankfurter grüne Soße          12           2   

        QuantityPerUnit  UnitPrice  UnitsInStock  UnitsOnOrder  ReorderLevel  \
0    10 boxes x 20 bags      18.00            39             0            10   
1    24 - 12 oz bottles      19.00 

In [21]:
#Print order detail table
order_detail = ex_query("""SELECT *
                      FROM orderdetail""")
print(order_detail)

              Id  OrderId  ProductId  UnitPrice  Quantity  Discount
0       10248/11    10248         11      14.00        12       0.0
1       10248/42    10248         42       9.80        10       0.0
2       10248/72    10248         72      34.80         5       0.0
3       10249/14    10249         14      18.60         9       0.0
4       10249/51    10249         51      42.40        40       0.0
...          ...      ...        ...        ...       ...       ...
621878  27065/22    27065         22      21.00        20       0.0
621879  27065/77    27065         77      13.00        11       0.0
621880  27065/17    27065         17      39.00        45       0.0
621881   27065/6    27065          6      25.00         7       0.0
621882  27065/65    27065         65      21.05        27       0.0

[621883 rows x 6 columns]


In [22]:
#Print two empty tables in our database
Customer_Customer_Demo = ex_query("""SELECT *
                      FROM customercustomerdemo""")
print(Customer_Customer_Demo)

customer_demographic = ex_query("""SELECT *
                      FROM customerdemographic""")
print(customer_demographic)

Empty DataFrame
Columns: [Id, CustomerTypeId]
Index: []
Empty DataFrame
Columns: [Id, CustomerDesc]
Index: []


In [23]:
#Print region table
region = ex_query("""SELECT *
                      FROM region""")
print(region)

   Id RegionDescription
0   1           Eastern
1   2           Western
2   3          Northern
3   4          Southern


In [24]:
#Print territory table
territory = ex_query("""SELECT *
                      FROM territory""")
print(territory)

       Id TerritoryDescription  RegionId
0   01581             Westboro         1
1   01730              Bedford         1
2   01833            Georgetow         1
3   02116               Boston         1
4   02139            Cambridge         1
..    ...                  ...       ...
48  95054          Santa Clara         2
49  95060           Santa Cruz         2
50  98004             Bellevue         2
51  98052              Redmond         2
52  98104              Seattle         2

[53 rows x 3 columns]


In [25]:
#Print employee territory table
employee_territory = ex_query("""SELECT *
                      FROM employeeterritory""")
print(employee_territory)

         Id  EmployeeId TerritoryId
0   1/06897           1       06897
1   1/19713           1       19713
2   2/01581           2       01581
3   2/01730           2       01730
4   2/01833           2       01833
5   2/02116           2       02116
6   2/02139           2       02139
7   2/02184           2       02184
8   2/40222           2       40222
9   3/30346           3       30346
10  3/31406           3       31406
11  3/32859           3       32859
12  3/33607           3       33607
13  4/20852           4       20852
14  4/27403           4       27403
15  4/27511           4       27511
16  5/02903           5       02903
17  5/07960           5       07960
18  5/08837           5       08837
19  5/10019           5       10019
20  5/10038           5       10038
21  5/11747           5       11747
22  5/14450           5       14450
23  6/85014           6       85014
24  6/85251           6       85251
25  6/98004           6       98004
26  6/98052           6     

## Data Exploration

We notice that there are two empty tables in our database. Customer Customer Demo and Customer Demographic are both empty. Now that we see what each table looks like, let's run some queries to find out a few things.

For instance what is the distribution among the different regions? How many territories are in each region?

In [26]:
#Count territories in each region
ex_query("""SELECT regiondescription, COUNT(*)
              FROM territory
             INNER JOIN region
                ON region.id = territory.regionid
             GROUP BY regionid""")

RegionDescription  COUNT(*)
0           Eastern        19
1           Western        15
2          Northern        11
3          Southern         8

The region with the most territories is the Eastern Region with 19. The Southern Region only has 8 territories. However this doesn't necessarily correlate to more sales. The company could have combined the regions in a way that all four regions have more or less the same amount of orders. To find this out we need to combine several tables.

In [27]:
#Calculate total sales and total units sold for each region
ex_query("""SELECT SUM(quantity) AS quantity, SUM(quantity * unitprice) AS total_sales, regiondescription
              FROM orderdetail
             INNER JOIN orders
                ON orders.id = orderdetail.orderid
             INNER JOIN employee
                ON employee.id = orders.employeeid
             INNER JOIN employeeterritory
                ON employeeterritory.employeeid = employee.id
             INNER JOIN territory
                ON territory.id = employeeterritory.territoryid
             INNER JOIN region
                ON region.id = territory.regionid
             GROUP BY RegionDescription
             ORDER BY quantity DESC"""
             )

quantity   total_sales RegionDescription
0  33262613  9.616455e+08           Eastern
1  26392010  7.635045e+08           Western
2  19198200  5.543704e+08          Northern
3   7400168  2.132334e+08          Southern

It is confirmed by the rank of quantity that the Eastern region has the most units sold and has the highest sales. Whatever employees have the Eastern region must have the highest sales.

Let's now see what cities are the most profitable. It's not necessarily true that all the cities in the Eastern region are most profitable. Let's find out.

In [28]:
# Total units sold and total sales by territory
ex_query("""SELECT SUM(quantity) AS quantity, SUM(quantity * unitprice) AS total_sales, territorydescription ,regiondescription
              FROM orderdetail
             INNER JOIN orders
                ON orders.id = orderdetail.orderid
             INNER JOIN employee
                ON employee.id = orders.employeeid
             INNER JOIN employeeterritory
                ON employeeterritory.employeeid = employee.id
             INNER JOIN territory
                ON territory.id = employeeterritory.territoryid
             INNER JOIN region
                ON region.id = territory.regionid
             GROUP BY territoryDescription
             ORDER BY total_sales DESC"""
             )

quantity   total_sales TerritoryDescription RegionDescription
0    3621402  1.044609e+08             New York           Eastern
1    1850042  5.330835e+07                Tampa          Southern
2    1850042  5.330835e+07             Savannah          Southern
3    1850042  5.330835e+07              Orlando          Southern
4    1850042  5.330835e+07              Atlanta          Southern
5    1810701  5.223046e+07           Providence           Eastern
6    1810701  5.223046e+07           Morristown           Eastern
7    1810701  5.223046e+07             Mellvile           Eastern
8    1810701  5.223046e+07             Fairport           Eastern
9    1810701  5.223046e+07               Edison           Eastern
10   1780500  5.152730e+07              Seattle           Western
11   1780500  5.152730e+07           Scottsdale           Western
12   1780500  5.152730e+07              Redmond           Western
13   1780500  5.152730e+07              Phoenix           Western
14   1780500  5.152730e+07             Bellevue           Western
15   1767409  5.083916e+07            Rockville           Eastern
16   1767409  5.083916e+07           Greensboro           Eastern
17   1767409  5.083916e+07                 Cary           Eastern
18   1761556  5.076263e+07                 Troy          Northern
19   1761556  5.076263e+07           Southfield          Northern
20   1761556  5.076263e+07            Roseville          Northern
21   1761556  5.076263e+07           Portsmouth          Northern
22   1761556  5.076263e+07          Minneapolis          Northern
23   1761556  5.076263e+07               Hollis          Northern
24   1761556  5.076263e+07     Bloomfield Hills          Northern
25   1748951  5.058680e+07         Santa Monica           Western
26   1748951  5.058680e+07           Santa Cruz           Western
27   1748951  5.058680e+07          Santa Clara           Western
28   1748951  5.058680e+07        San Francisco           Western
29   1748951  5.058680e+07           Menlo Park           Western
30   1748951  5.058680e+07      Hoffman Estates           Western
31   1748951  5.058680e+07               Denver           Western
32   1748951  5.058680e+07     Colorado Springs           Western
33   1748951  5.058680e+07              Chicago           Western
34   1748951  5.058680e+07             Campbell           Western
35   1730242  4.981720e+07               Wilton           Eastern
36   1730242  4.981720e+07               Neward           Eastern
37   1716827  4.975801e+07               Racine          Northern
38   1716827  4.975801e+07         Philadelphia          Northern
39   1716827  4.975801e+07              Findlay          Northern
40   1716827  4.975801e+07            Beachwood          Northern
41   1689285  4.912578e+07             Westboro           Eastern
42   1689285  4.912578e+07           Louisville           Eastern
43   1689285  4.912578e+07            Georgetow           Eastern
44   1689285  4.912578e+07            Cambridge           Eastern
45   1689285  4.912578e+07            Braintree           Eastern
46   1689285  4.912578e+07               Boston           Eastern
47   1689285  4.912578e+07              Bedford           Eastern

As we can see four of the top five best selling cities are in the Southern region. This may be a reason why the Southern region has so few territories.

Moving on let's look at some of our other tables and let's do so using a window. I'd like to know the average sales by city but I'd like to retain the rows instead of collapsing them in order to see if there are a lot of high values and low values or values close to the average. In a way, I'd like to eyeball the standard deviation.

In [29]:
#Using a window to see average sales per region and each individual sale in that region
ex_query("""SELECT territory.id, AVG(quantity * unitprice) OVER (Partition BY territorydescription) AS avg_sales, territorydescription , quantity * unitprice AS sale
              FROM orderdetail
             INNER JOIN orders
                ON orders.id = orderdetail.orderid
             INNER JOIN employee
                ON employee.id = orders.employeeid
             INNER JOIN employeeterritory
                ON employeeterritory.employeeid = employee.id
             INNER JOIN territory
                ON territory.id = employeeterritory.territoryid
             INNER JOIN region
                ON region.id = territory.regionid
             WHERE TerritoryDescription = 'New York'
             ORDER BY avg_sales DESC"""
             )

Id   avg_sales TerritoryDescription    sale
0       10019  734.161581             New York  168.00
1       10038  734.161581             New York  168.00
2       10019  734.161581             New York   98.00
3       10038  734.161581             New York   98.00
4       10019  734.161581             New York  174.00
...       ...         ...                  ...     ...
142281  10038  734.161581             New York  897.00
142282  10019  734.161581             New York  522.75
142283  10038  734.161581             New York  522.75
142284  10019  734.161581             New York  374.76
142285  10038  734.161581             New York  374.76

[142286 rows x 4 columns]

As we can see the sales for New York can be very close to the avg or very far. Let's find the max and min sale for New York.

In [30]:
# Using window and Common Table Expression to find max and min sale for New York territory
ex_query("""WITH
            orig AS (SELECT orders.id, AVG(quantity * unitprice) OVER (Partition BY territorydescription) AS avg_sales, territorydescription , quantity * unitprice AS sale
              FROM orderdetail
             INNER JOIN orders
                ON orders.id = orderdetail.orderid
             INNER JOIN employee
                ON employee.id = orders.employeeid
             INNER JOIN employeeterritory
                ON employeeterritory.employeeid = employee.id
             INNER JOIN territory
                ON territory.id = employeeterritory.territoryid
             INNER JOIN region
                ON region.id = territory.regionid
             WHERE TerritoryDescription = 'New York'
             ORDER BY avg_sales DESC)

     SELECT *
     FROM orig
     WHERE sale = (SELECT MAX(sale) FROM orig)
     
      
     UNION

     SELECT *
     FROM orig
     WHERE sale = (SELECT MIN(sale) FROM orig)"""
             )

id   avg_sales territorydescription     sale
0   11642  734.161581             New York  13175.0
1   13419  734.161581             New York  13175.0
2   13508  734.161581             New York      2.5
3   13723  734.161581             New York  13175.0
4   14416  734.161581             New York      2.5
5   14622  734.161581             New York      2.5
6   14720  734.161581             New York      2.5
7   15767  734.161581             New York  13175.0
8   16325  734.161581             New York  13175.0
9   16558  734.161581             New York      2.5
10  16567  734.161581             New York  13175.0
11  16659  734.161581             New York  13175.0
12  16835  734.161581             New York      2.5
13  17829  734.161581             New York      2.5
14  18367  734.161581             New York  13175.0
15  18497  734.161581             New York      2.5
16  18513  734.161581             New York  13175.0
17  19205  734.161581             New York  13175.0
18  19256  734.161581             New York      2.5
19  19416  734.161581             New York  13175.0
20  19614  734.161581             New York  13175.0
21  19707  734.161581             New York      2.5
22  20449  734.161581             New York      2.5
23  21486  734.161581             New York  13175.0
24  22378  734.161581             New York  13175.0
25  23000  734.161581             New York      2.5
26  23053  734.161581             New York  13175.0
27  23580  734.161581             New York      2.5
28  25061  734.161581             New York      2.5
29  25461  734.161581             New York      2.5
30  26472  734.161581             New York      2.5
31  26920  734.161581             New York  13175.0
32  27061  734.161581             New York      2.5

We seem to have many different order IDs with the same max and min sale values. We have to remember that this is a sample and fictious database. Regardless of this fact we can still find out some interesting things about this database. Let's take a look at some other tables.

Let's find the maxiumum unit prices for the different products they import and export. We can take advantage of the fact that SQL will list the first five rows and the last five rows and order our results properly.

In [31]:
#Look at product table with lowest price and highest price.
ex_query("""SELECT *
              FROM product
             ORDER BY unitprice""")

Id              ProductName  SupplierId  CategoryId       QuantityPerUnit  \
0   33                  Geitost          15           4                 500 g   
1   24       Guaraná Fantástica          10           1      12 - 355 ml cans   
2   13                    Konbu           6           8              2 kg box   
3   52                 Filo Mix          24           5       16 - 2 kg boxes   
4   54                Tourtière          25           6               16 pies   
..  ..                      ...         ...         ...                   ...   
72  18         Carnarvon Tigers           7           8            16 kg pkg.   
73  20   Sir Rodney's Marmalade           8           3         30 gift boxes   
74   9          Mishi Kobe Niku           4           6      18 - 500 g pkgs.   
75  29  Thüringer Rostbratwurst          12           6  50 bags x 30 sausgs.   
76  38            Côte de Blaye          18           1    12 - 75 cl bottles   

    UnitPrice  UnitsInStock  UnitsOnOrder  ReorderLevel  Discontinued  
0        2.50           112             0            20             0  
1        4.50            20             0             0             1  
2        6.00            24             0             5             0  
3        7.00            38             0            25             0  
4        7.45            21             0            10             0  
..        ...           ...           ...           ...           ...  
72      62.50            42             0             0             0  
73      81.00            40             0             0             0  
74      97.00            29             0             0             1  
75     123.79             0             0             0             1  
76     263.50            17             0            15             0  

[77 rows x 10 columns]

Their cheapest product is 2.50 and their most expensive product is 263.50. I'm curious now, how many products have been discontinued and where those products come from? Is there a country that Northwind is phasing out and not considering working with anymore?

First, let's find how many products are discontinued.

In [32]:
#Finding what products are discontinued, using DISTINCT function
ex_query("""SELECT DISTINCT(Productname), id
              FROM product
             WHERE discontinued = 1
             ORDER BY productname""")

ProductName  Id
0                   Alice Mutton  17
1         Chef Anton's Gumbo Mix   5
2             Guaraná Fantástica  24
3                Mishi Kobe Niku   9
4                  Perth Pasties  53
5              Rössle Sauerkraut  28
6  Singaporean Hokkien Fried Mee  42
7        Thüringer Rostbratwurst  29

So there are 8 discontinued products. Let's see where each of these products are from.

In [33]:
#Find country of supplier
ex_query(""" SELECT productname, supplier.country, supplier.city, unitprice
              FROM product
              LEFT JOIN supplier
                ON supplier.id = product.supplierid
             WHERE discontinued = 1
             ORDER BY productname""")

ProductName    Country         City  UnitPrice
0                   Alice Mutton  Australia    Melbourne      39.00
1         Chef Anton's Gumbo Mix        USA  New Orleans      21.35
2             Guaraná Fantástica     Brazil    Sao Paulo       4.50
3                Mishi Kobe Niku      Japan        Tokyo      97.00
4                  Perth Pasties  Australia       Sydney      32.80
5              Rössle Sauerkraut    Germany    Frankfurt      45.60
6  Singaporean Hokkien Fried Mee  Singapore    Singapore      14.00
7        Thüringer Rostbratwurst    Germany    Frankfurt     123.79

Other than two products from Germany and two products from Australia, there doesn't seem to be a trend for discontinued products. Let's check sales, what's in stock and orders.

In [34]:
# Computes unit price, units in stock, units on order, total quantity of units sold and total sales for discontinued products
ex_query("""SELECT productname, product.unitprice, MIN(unitsinstock) AS current_units_in_stock, SUM(unitsonorder) AS total_units_on_order, SUM(quantity) AS units_sold, SUM(orderdetail.unitprice * quantity) AS sale
              FROM product
              LEFT JOIN orderdetail
                ON orderdetail.productid = product.id
             WHERE discontinued = 1
             GROUP BY productname
             ORDER BY sale DESC""")

ProductName  UnitPrice  current_units_in_stock  \
0        Thüringer Rostbratwurst     123.79                       0   
1                Mishi Kobe Niku      97.00                      29   
2              Rössle Sauerkraut      45.60                      26   
3                   Alice Mutton      39.00                       0   
4                  Perth Pasties      32.80                       0   
5         Chef Anton's Gumbo Mix      21.35                       0   
6  Singaporean Hokkien Fried Mee      14.00                      26   
7             Guaraná Fantástica       4.50                      20   

   total_units_on_order  units_sold         sale  
0                     0      208449  25799290.77  
1                     0      205173  19901393.00  
2                     0      204781   9335695.20  
3                     0      204220   7961920.20  
4                     0      207113   6791135.00  
5                     0      206906   4416881.95  
6                     0      203255   2845144.40  
7                     0      203606    915947.10

We can see that this fictious database is smart enough not to have units on order of products they have already discontinued. Also, based on the units in stock, we can surmise that four companies decided it wasn't worth it for them. Four other companies that still have units in stock for the last order, might just not have been able to sell enough. Let's see how these discontinued products ranked among other products and see if we can either disprove or corroborate that theory.

In [35]:
#Top 30 selling items for Northwind Trades
ex_query("""SELECT productname, product.unitprice, SUM(unitsinstock), SUM(unitsonorder), SUM(quantity), SUM(orderdetail.unitprice * quantity) AS sale, discontinued
              FROM product
              LEFT JOIN orderdetail
                ON orderdetail.productid = product.id
             GROUP BY productname
             ORDER BY sale DESC
             LIMIT 30""")

ProductName  UnitPrice  SUM(unitsinstock)  \
0                     Côte de Blaye     263.50             137632   
1           Thüringer Rostbratwurst     123.79                  0   
2                   Mishi Kobe Niku      97.00             233682   
3            Sir Rodney's Marmalade      81.00             323600   
4                  Carnarvon Tigers      62.50             339318   
5              Raclette Courdavault      55.00             641717   
6             Manjimup Dried Apples      53.00             163580   
7                    Tarte au sucre      49.30             137632   
8                       Ipoh Coffee      46.00             138244   
9                 Rössle Sauerkraut      45.60             209638   
10                     Vegie-spread      43.90             194400   
11               Schoggi Schokolade      43.90             395185   
12       Northwoods Cranberry Sauce      40.00              48006   
13                     Alice Mutton      39.00                  0   
14        Queso Manchego La Pastora      38.00             692386   
15           Gnocchi di nonna Alice      38.00             170058   
16                 Gudbrandsdalsost      36.00             209560   
17           Mozzarella di Giovanni      34.80             112252   
18                Camembert Pierrot      34.00             153634   
19        Wimmers gute Semmelknödel      33.25             177320   
20                    Perth Pasties      32.80                  0   
21               Mascarpone Fabioli      32.00              72162   
22              Gumbär Gummibärchen      31.23             121305   
23                            Ikura      31.00             248341   
24  Uncle Bob's Organic Dried Pears      30.00             120885   
25                   Sirop d'érable      28.50             906825   
26            Nord-Ost Matjeshering      25.89              81040   
27                       Gravad lax      26.00              88495   
28     Grandma's Boysenberry Spread      25.00             964680   
29                     Pâté chinois      24.00             929200   

    SUM(unitsonorder)  SUM(quantity)         sale  Discontinued  
0                   0         205055  54017816.20             0  
1                   0         208449  25799290.77             1  
2                   0         205173  19901393.00             1  
3                   0         207647  16817689.80             0  
4                   0         207126  12943675.00             0  
5                   0         206729  11364111.00             0  
6                   0         209216  11086232.60             0  
7                   0         206720  10187732.00             0  
8               81320         207646   9550115.20             0  
9                   0         204781   9335695.20             1  
10                  0         209875   9211673.30             0  
11                  0         205553   9022984.70             0  
12                  0         203223   8127800.00             0  
13                  0         204220   7961920.20             1  
14                  0         205986   7827262.80             0  
15              80980         205532   7807343.20             0  
16                  0         208439   7502407.20             0  
17                  0         205172   7137675.60             0  
18                  0         207652   7056836.00             0  
19             644800         206594   6867654.50             0  
20                  0         207113   6791135.00             1  
21             320720         205060   6561587.20             0  
22                  0         204972   6399294.27             0  
23                  0         203658   6312536.20             0  
24                  0         205437   6162684.00             0  
25                  0         205800   5864553.30             0  
26                  0         206411   5342911.65             0  
27             402250         202955

One thing that jumps out immediately is the direct linear relationship between price per unit and sales. The higher the price, the more money was made. All the companies sold around the same amount of units, give or take a couple thousand units. We also see that two of the three top selling products were discontinued. We can see that it wasn't due to lack of sales. Mishi Kobe Niku disproves the theory of not enough sales being correlated with still having units in stock. Perhaps the products were too heavy and the cost of freight was too much to make it worth it for the customers. Let's see if we can find that out.

In [36]:
# Calculate total freight cost versus total sales
ex_query("""SELECT productname, SUM(orderdetail.unitprice * quantity) AS total_sales, SUM(freight) AS Freight_Cost
              FROM product
              LEFT JOIN orderdetail
                ON orderdetail.productid = product.id
              LEFT JOIN orders
                ON orders.id = orderdetail.orderid
             LEFT JOIN customer
                ON customer.id = orders.customerid
             WHERE discontinued = 1
             GROUP BY productname
             ORDER BY total_sales DESC
             LIMIT 30""")

ProductName  total_sales  Freight_Cost
0        Thüringer Rostbratwurst  25799290.77    2737401.00
1                Mishi Kobe Niku  19901393.00    2704939.75
2              Rössle Sauerkraut   9335695.20    2714471.25
3                   Alice Mutton   7961920.20    2715679.75
4                  Perth Pasties   6791135.00    2716876.25
5         Chef Anton's Gumbo Mix   4416881.95    2712420.50
6  Singaporean Hokkien Fried Mee   2845144.40    2700736.00
7             Guaraná Fantástica    915947.10    2701958.25

From the documentation I found on the database, the freight column is a decimal column which normally denotes currency in this case, therefore this must be how much it cost for freight. Let's verify if there are other columns that have decimal type and if so, should they be decimal types.

In [37]:
#Variable types of orderdetail table
ex_query('PRAGMA table_info(orderdetail)')

cid       name           type  notnull dflt_value  pk
0    0         Id  VARCHAR(8000)        0       None   1
1    1    OrderId        INTEGER        1       None   0
2    2  ProductId        INTEGER        1       None   0
3    3  UnitPrice        DECIMAL        1       None   0
4    4   Quantity        INTEGER        1       None   0
5    5   Discount         DOUBLE        1       None   0

In [38]:
#Variable types of orders table
ex_query('PRAGMA table_info(orders)')

cid            name           type  notnull dflt_value  pk
0     0              Id        INTEGER        0       None   1
1     1      CustomerId  VARCHAR(8000)        0       None   0
2     2      EmployeeId        INTEGER        1       None   0
3     3       OrderDate  VARCHAR(8000)        0       None   0
4     4    RequiredDate  VARCHAR(8000)        0       None   0
5     5     ShippedDate  VARCHAR(8000)        0       None   0
6     6         ShipVia        INTEGER        0       None   0
7     7         Freight        DECIMAL        1       None   0
8     8        ShipName  VARCHAR(8000)        0       None   0
9     9     ShipAddress  VARCHAR(8000)        0       None   0
10   10        ShipCity  VARCHAR(8000)        0       None   0
11   11      ShipRegion  VARCHAR(8000)        0       None   0
12   12  ShipPostalCode  VARCHAR(8000)        0       None   0
13   13     ShipCountry  VARCHAR(8000)        0       None   0

In [39]:
#Variable types of product table
ex_query('PRAGMA table_info(product)')

cid             name           type  notnull dflt_value  pk
0    0               Id        INTEGER        0       None   1
1    1      ProductName  VARCHAR(8000)        0       None   0
2    2       SupplierId        INTEGER        1       None   0
3    3       CategoryId        INTEGER        1       None   0
4    4  QuantityPerUnit  VARCHAR(8000)        0       None   0
5    5        UnitPrice        DECIMAL        1       None   0
6    6     UnitsInStock        INTEGER        1       None   0
7    7     UnitsOnOrder        INTEGER        1       None   0
8    8     ReorderLevel        INTEGER        1       None   0
9    9     Discontinued        INTEGER        1       None   0

Looking through the table schema, I found only three columns in our database with decimal types and two of them are clearly prices. We should safely be able to assume that this is also a price of the freight and not it's weight. Let's remind ourselves of the freight and sales columns.

In [40]:
#Total sales vs. total freight cost vs. profit of discontinued products
ex_query("""SELECT productname, SUM(orderdetail.unitprice * quantity) AS sale, SUM(freight) as freight, SUM(orderdetail.unitprice * quantity) - SUM(freight)  AS profit
              FROM product
              LEFT JOIN orderdetail
                ON orderdetail.productid = product.id
              LEFT JOIN orders
                ON orders.id = orderdetail.orderid
             LEFT JOIN customer
                ON customer.id = orders.customerid
             WHERE discontinued = 1
             GROUP BY productname
             ORDER BY profit DESC
             LIMIT 30""")

ProductName         sale     freight       profit
0        Thüringer Rostbratwurst  25799290.77  2737401.00  23061889.77
1                Mishi Kobe Niku  19901393.00  2704939.75  17196453.25
2              Rössle Sauerkraut   9335695.20  2714471.25   6621223.95
3                   Alice Mutton   7961920.20  2715679.75   5246240.45
4                  Perth Pasties   6791135.00  2716876.25   4074258.75
5         Chef Anton's Gumbo Mix   4416881.95  2712420.50   1704461.45
6  Singaporean Hokkien Fried Mee   2845144.40  2700736.00    144408.40
7             Guaraná Fantástica    915947.10  2701958.25  -1786011.15

While we don't know what amount of profit each company needs to make it worth for them, we know that it can't be negative. They can't pay employees negative currency. We could probably surmise that the bottom three stopped either producing or working with Northwind Trades due to low profits.

Now let's see if the suppliers of the other products, might just be too far from the customers to make it a worthwhile endeavor either for the suppliers or the customers. Perishables like pastries might not be realistic to ship if the distance is too far, at least in our fictious database.

In [41]:
#Display customer and supplier countries
pd.set_option('display.max_rows', None)
ex_query("""SELECT DISTINCT productname, customer.country AS customer_country, supplier.country AS supplier_country
              FROM product
              LEFT JOIN supplier
                ON supplier.id = product.supplierid
              LEFT JOIN orderdetail
                ON orderdetail.productid = product.id
              LEFT JOIN orders
                ON orders.id = orderdetail.orderid
             LEFT JOIN customer
                ON customer.id = orders.customerid
             WHERE discontinued = 1
             """)

ProductName customer_country supplier_country
0           Chef Anton's Gumbo Mix          Austria              USA
1           Chef Anton's Gumbo Mix              USA              USA
2           Chef Anton's Gumbo Mix           Brazil              USA
3           Chef Anton's Gumbo Mix            Italy              USA
4           Chef Anton's Gumbo Mix               UK              USA
5           Chef Anton's Gumbo Mix             None              USA
6           Chef Anton's Gumbo Mix          Germany              USA
7           Chef Anton's Gumbo Mix           Norway              USA
8           Chef Anton's Gumbo Mix        Argentina              USA
9           Chef Anton's Gumbo Mix           France              USA
10          Chef Anton's Gumbo Mix         Portugal              USA
11          Chef Anton's Gumbo Mix      Switzerland              USA
12          Chef Anton's Gumbo Mix           Canada              USA
13          Chef Anton's Gumbo Mix           Sweden              USA
14          Chef Anton's Gumbo Mix           Mexico              USA
15          Chef Anton's Gumbo Mix          Denmark              USA
16          Chef Anton's Gumbo Mix          Finland              USA
17          Chef Anton's Gumbo Mix        Venezuela              USA
18          Chef Anton's Gumbo Mix            Spain              USA
19          Chef Anton's Gumbo Mix           Poland              USA
20          Chef Anton's Gumbo Mix          Belgium              USA
21          Chef Anton's Gumbo Mix          Ireland              USA
22                 Mishi Kobe Niku           Brazil            Japan
23                 Mishi Kobe Niku          Germany            Japan
24                 Mishi Kobe Niku          Ireland            Japan
25                 Mishi Kobe Niku              USA            Japan
26                 Mishi Kobe Niku               UK            Japan
27                 Mishi Kobe Niku           Norway            Japan
28                 Mishi Kobe Niku           Canada            Japan
29                 Mishi Kobe Niku        Argentina            Japan
30                 Mishi Kobe Niku          Belgium            Japan
31                 Mishi Kobe Niku           France            Japan
32                 Mishi Kobe Niku         Portugal            Japan
33                 Mishi Kobe Niku            Italy            Japan
34                 Mishi Kobe Niku          Finland            Japan
35                 Mishi Kobe Niku           Sweden            Japan
36                 Mishi Kobe Niku           Mexico            Japan
37                 Mishi Kobe Niku            Spain            Japan
38                 Mishi Kobe Niku          Denmark            Japan
39                 Mishi Kobe Niku        Venezuela            Japan
40                 Mishi Kobe Niku          Austria            Japan
41                 Mishi Kobe Niku      Switzerland            Japan
42                 Mishi Kobe Niku           Poland            Japan
43                    Alice Mutton           France        Australia
44                    Alice Mutton          Germany        Australia
45                    Alice Mutton              USA        Australia
46                    Alice Mutton          Belgium        Australia
47                    Alice Mutton           Mexico        Australia
48                    Alice Mutton           Canada        Australia
49                    Alice Mutton          Austria        Australia
50                    Alice Mutton           Sweden        Australia
51                    Alice Mutton               UK        Australia
52                    Alice Mutton            Spain        Australia
53                    Alice Mutton             None        Australia
54                    Alice Mutton            Italy        Australia
55                    Alice Mutton           Brazil        Australia
56                    Alice Mutton           Norway        Australia
57                   

It appears that each of the companies ships to all of the same places. It would be nice to verify that by listing the number of countries that each company ships to. Let's see if we can do that.

In [42]:
#Counts the number of countries that each supplier of discontinued products shipped to
ex_query("""WITH 
            pcs AS (SELECT DISTINCT productname, customer.country AS customer_country, supplier.country AS supplier_country
              FROM product
              LEFT JOIN supplier
                ON supplier.id = product.supplierid
              LEFT JOIN orderdetail
                ON orderdetail.productid = product.id
              LEFT JOIN orders
                ON orders.id = orderdetail.orderid
             LEFT JOIN customer
                ON customer.id = orders.customerid
             WHERE discontinued = 1)
             
             SELECT COUNT(DISTINCT pcs.customer_country)
               FROM pcs
              GROUP BY productname
             """)

COUNT(DISTINCT pcs.customer_country)
0                                    21
1                                    21
2                                    21
3                                    21
4                                    21
5                                    21
6                                    21
7                                    21

Each company ships to the same 21 countries. This means that the Northwind Trades would typically dictate where the products go. Since all the companies have similar freight costs and ship to all the same countries, it doesn't seem that distance would be a reason why the products were discontinued. There may not be data in our database that can tell us why the products were discontinued, it is after all fictitious.

Let's take a look at some other tables. Looking at the shipper table there are 3 shipping companies used by Northwind Trades, let's see which of those is the most frequently used.

In [43]:
# 1 - Speedy Express, 2 - United Package, 3- Federal Shipping
# Total orders for three shipping companies
ex_query("""SELECT COUNT(shipvia) AS total_orders, shipvia
              FROM orders
             GROUP BY shipvia
             ORDER BY shipvia DESC""")

total_orders  ShipVia
0          5654        3
1          5592        2
2          5572        1

Federal Shipping is used the most while Speedy Express is used the least. Relatively there's not that much difference between them. But does speedy express really ship the fastest? Who ships the fastest on average? Given that all the companies ship to the same 21 countries, let's see who is the fastest.

In [44]:
#Calculates average shipping time for each of the three companies
ex_query("""SELECT AVG((Julianday(Shippeddate)  - Julianday(Orderdate))) AS shipping_time, shipvia
              FROM orders
             GROUP BY shipvia
             ORDER BY shipping_time""")

shipping_time  ShipVia
0       7.794423        1
1       7.878872        3
2       8.000029        2

All the shipping times are on average are the same but Speedy Express is the fastest while United Package is the slowest. However there's not enough of a difference. The difference between the fastest and slowest company is about 5 hours. Given that the average time to ship is about 8 days, it's not a big enough difference to choose one company or the other. One missing piece of information would be the time for some of the rows. Arriving in the evening or in the morning could throw off the averages completely. Below is a table that shows the JulianDay function was working correctly.

In [45]:
#Verifies JulianDay function worked correctly by showing difference in time from order date to shipped date by row
pd.set_option('display.max_rows', 25)
ex_query("""SELECT Shippeddate, Orderdate, Julianday(Shippeddate) - Julianday(Orderdate) AS diff
              FROM orders
           """)

ShippedDate            OrderDate       diff
0               2012-07-16           2012-07-04  12.000000
1               2012-07-10           2012-07-05   5.000000
2               2012-07-12           2012-07-08   4.000000
3               2012-07-15           2012-07-08   7.000000
4               2012-07-11           2012-07-09   2.000000
...                    ...                  ...        ...
16813  2013-07-02 16:05:51  2013-06-29 21:05:55   2.791620
16814  2014-01-27 02:14:31  2014-01-19 12:27:11   7.574537
16815  2014-10-16 06:26:55  2014-10-15 09:51:09   0.858171
16816  2013-02-24 10:15:47  2013-02-07 02:06:05  17.340069
16817  2013-09-03 14:09:08  2013-08-31 02:59:28   3.465046

[16818 rows x 3 columns]

Looking at the tables above the products that Northwind Trades imports and exports fall into eight categories. Let's pose a bit more of a difficult question. What is the most popular category sought by customers whose company name starts with the letter A? To answer that question, we'll have to combine a few tables and while we're at it why not create a view.

In [46]:
#Creates view of company names that start with A and the most popular products they buy
pd.set_option('display.max_rows', None)
# ex_query("""CREATE VIEW popular AS
#            SELECT companyname, categoryname, COUNT(*) AS category_count
#              FROM customer
#              LEFT JOIN orders
#                ON orders.customerid = customer.id
#              LEFT JOIN orderdetail
#                ON orderdetail.orderid = orders.id
#              LEFT JOIN product
#               ON product.id = orderdetail.productid
#              LEFT JOIN category
#                ON category.id = product.categoryid
#             WHERE companyname LIKE 'A%'
#             GROUP BY companyname, categoryname
#             ORDER BY companyname, category_count DESC""")

In [47]:
# Displays the view created above
ex_query("""SELECT companyname, categoryname, MAX(category_count) AS most_popular_category
              FROM popular
             GROUP BY companyname""")

CompanyName CategoryName  most_popular_category
0                 Alfreds Futterkiste  Confections                   1261
1  Ana Trujillo Emparedados y helados  Confections                   1223
2             Antonio Moreno Taquería  Confections                   1222
3                     Around the Horn  Confections                    990

Among the companies that start with A, it is a consensus that they order confections the most. That's the most common category among companies that start with A but now I'm curious what actually the most common category is amongst all customers. In addition we will rank them because SQL allows us to.

In [48]:
#Ranks most popular categories of products ordered by customers
ex_query("""SELECT  categoryname, COUNT(*) AS category_count, RANK() OVER(ORDER BY COUNT(*) DESC) AS Rank
              FROM customer
              LEFT JOIN orders
                ON orders.customerid = customer.id
              LEFT JOIN orderdetail
                ON orderdetail.orderid = orders.id
              LEFT JOIN product
                ON product.id = orderdetail.productid
              LEFT JOIN category
                ON category.id = product.categoryid
             GROUP BY categoryname
             ORDER BY  category_count DESC""")

CategoryName  category_count  Rank
0     Confections          104963     1
1       Beverages           97033     2
2      Condiments           96807     3
3         Seafood           96766     4
4  Dairy Products           80737     5
5  Grains/Cereals           56459     6
6    Meat/Poultry           48540     7
7         Produce           40501     8

Confections are the most popular among Northwind Trades customers as a whole. Looking through some of the customer company names, it was mostly the most popular among individual customers.

Now let's look at the employee table. I would like to know who was the youngest when they joined the company. I'd also like their first and last name in one column with, the hire date, birth date and how old they were in years when they joined the company.

In [49]:
#Calculates age when hired of all employees
ex_query("""SELECT firstname || ' ' || lastname AS name, hiredate, birthdate, ROUND(((julianday(hiredate) - julianday(birthdate))/365),1) AS age_when_hired
              FROM employee
             ORDER BY age_when_hired""")

name    HireDate   BirthDate  age_when_hired
0   Janet Leverling  2024-04-01  1995-08-30            28.6
1    Anne Dodsworth  2026-11-15  1998-01-27            28.8
2    Michael Suyama  2025-10-17  1995-07-02            30.3
3       Robert King  2026-01-02  1992-05-29            33.6
4    Laura Callahan  2026-03-05  1990-01-09            36.2
5   Steven Buchanan  2025-10-17  1987-03-04            38.6
6     Andrew Fuller  2024-08-14  1984-02-19            40.5
7     Nancy Davolio  2024-05-01  1980-12-08            43.4
8  Margaret Peacock  2025-05-03  1969-09-19            55.7

We have some new information that I was not previously aware of. This database is from the future. All hire dates are in 2024 or later. The youngest person in the company is not the person who was the youngest when they were hired. Janet Leverling was the youngest at 28.6 years old when she was hired.

Does seniority mean that an employee will have more responsibility? Are things left handled by younger or older employees? Let's see.

In [50]:
# Calculates seniority, age and total order ranks
ex_query("""SELECT orders.employeeid,
                   firstname || ' ' || lastname AS name, 
                   hiredate, 
                   RANK() OVER(ORDER BY hiredate) AS Seniority, 
                   SUM(orders.id) AS total_orders, 
                   RANK() OVER(ORDER BY SUM(orders.id) DESC) AS total_order_rank, 
                   birthdate, 
                   RANK() OVER(ORDER BY birthdate) AS oldest
              FROM employee
              LEFT JOIN orders
                ON orders.employeeid = employee.id
             GROUP BY employee.id""")

EmployeeId              name    HireDate  Seniority  total_orders  \
0           3   Janet Leverling  2024-04-01          1      36206952   
1           1     Nancy Davolio  2024-05-01          2      35590617   
2           2     Andrew Fuller  2024-08-14          3      33819042   
3           4  Margaret Peacock  2025-05-03          4      35465339   
4           5   Steven Buchanan  2025-10-17          5      34831566   
5           6    Michael Suyama  2025-10-17          5      34557586   
6           7       Robert King  2026-01-02          7      34306453   
7           8    Laura Callahan  2026-03-05          8      34000177   
8           9    Anne Dodsworth  2026-11-15          9      34987285   

   total_order_rank   BirthDate  oldest  
0                 1  1995-08-30       8  
1                 2  1980-12-08       2  
2                 9  1984-02-19       3  
3                 3  1969-09-19       1  
4                 5  1987-03-04       4  
5                 6  1995-07-02       7  
6                 7  1992-05-29       6  
7                 8  1990-01-09       5  
8                 4  1998-01-27       9

It appears that seniority does matter. With the exception of two employees, Anne Dodsworth and Andrew Fuller, all the employees rank the highest in sales in terms of their seniority. It also appears as well that the older employees tend to do more sales. This shows that Northwind Trades functions like most companies function.

Let's now take a look at what is the biggest order for each of the nine employees. The best way to do this is to utilize the HAVING clause. Let's take a look at how that's done.

In [51]:
#Calculates Max Sale for each employee
ex_query("""SELECT orders.employeeid,
                   firstname || ' ' || lastname AS name,
                   (unitprice * quantity) AS sale
              FROM employee
              LEFT JOIN orders
                ON orders.employeeid = employee.id
              LEFT JOIN orderdetail
                ON orderdetail.orderid = orders.id
             GROUP BY orders.employeeid
            HAVING sale = MAX(sale)""")

EmployeeId              name     sale
0           1     Nancy Davolio  15810.0
1           2     Andrew Fuller  15810.0
2           3   Janet Leverling  13175.0
3           4  Margaret Peacock  13175.0
4           5   Steven Buchanan  13175.0
5           6    Michael Suyama  13175.0
6           7       Robert King  13175.0
7           8    Laura Callahan  13175.0
8           9    Anne Dodsworth  13175.0

## Conclusion

There are quite a few interesting takeaways to be had by pouring over the database of Northwind Trades.

- For ease of analysis, multiple, even round maximum and minimum values were chose to be filled in the database. It made looking at results easier and clear. The creators of the database didn't try to do too much. The same 21 countries were chosen for all products to be shipped to.

- Discontinued products generally had no clear reason in the data as to why they were discontinued. Three of the products just were clearly not profitable enough, with one company obtaining negative profits.

- Seniority in the company matters and they trust the older employees.


This has been an interesting look into a fictional database. I hope you enjoyed reading this project as much as I enjoyed doing it.